From 74379abe9036ac2ab082e920fd76314a6bd57168 Mon Sep 17 00:00:00 2001
From: Jouni Malinen <quic_jouni@quicinc.com>
Date: Wed, 11 Oct 2023 12:46:23 +0300
Subject: [PATCH] Allow RSNE Override element to override RSNE contents during
 parsing

Signed-off-by: Jouni Malinen <quic_jouni@quicinc.com>
---
 src/common/wpa_common.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/src/common/wpa_common.c b/src/common/wpa_common.c
index c858f0a77..1e66aa561 100644
--- a/src/common/wpa_common.c
+++ b/src/common/wpa_common.c
@@ -1888,6 +1888,14 @@ int wpa_parse_wpa_ie_rsn(const u8 *rsn_ie, size_t rsn_ie_len,
 		data->has_group = 1;
 		data->key_mgmt = WPA_KEY_MGMT_OSEN;
 		data->proto = WPA_PROTO_OSEN;
+	} else if (rsn_ie_len >= 2 + 4 + 2 && rsn_ie[1] >= 4 + 2 &&
+		   rsn_ie[1] == rsn_ie_len - 2 &&
+		   (WPA_GET_BE32(&rsn_ie[2]) == RSNE_OVERRIDE_IE_VENDOR_TYPE ||
+		    WPA_GET_BE32(&rsn_ie[2]) ==
+		    RSNE_OVERRIDE_2_IE_VENDOR_TYPE) &&
+		   WPA_GET_LE16(&rsn_ie[2 + 4]) == RSN_VERSION) {
+		pos = rsn_ie + 2 + 4 + 2;
+		left = rsn_ie_len - 2 - 4 - 2;
 	} else {
 		const struct rsn_ie_hdr *hdr;
 
-- 
2.43.0

