From 40b04b7034f4dd904fba27fa165a5146055c5fb5 Mon Sep 17 00:00:00 2001
From: Jouni Malinen <j@w1.fi>
Date: Wed, 6 Dec 2023 12:07:11 +0200
Subject: [PATCH] Document Tunnel-Password encoding for passphrase/PSK

Signed-off-by: Jouni Malinen <j@w1.fi>
---
 hostapd/hostapd.conf | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/hostapd/hostapd.conf b/hostapd/hostapd.conf
index 4db61d636..99da3db3f 100644
--- a/hostapd/hostapd.conf
+++ b/hostapd/hostapd.conf
@@ -1780,6 +1780,10 @@ own_ip_addr=127.0.0.1
 #	Tunnel-Password
 # 3 = ask RADIUS server during 4-way handshake if there is no locally
 #	configured PSK/passphrase for the STA
+#
+# The Tunnel-Password attribute in Access-Accept can contain either the
+# 8..63 character ASCII passphrase or a 64 hex character encoding of the PSK.
+#
 #wpa_psk_radius=0
 
 # Set of accepted key management algorithms (WPA-PSK, WPA-EAP, or both). The
-- 
2.43.0

