From 3978f7b92291a3199c2643a1952957e3e428fe74 Mon Sep 17 00:00:00 2001
From: Jurijs Soloveckis <jsoloveckis@maxlinear.com>
Date: Fri, 9 Jun 2023 09:25:05 +0200
Subject: [PATCH] WLANRTSYS-55993: Implement support for WHM: part1

---
 hostapd/config_file.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/hostapd/config_file.c b/hostapd/config_file.c
index 4f6339d91..af63da701 100644
--- a/hostapd/config_file.c
+++ b/hostapd/config_file.c
@@ -5705,6 +5705,8 @@ static int hostapd_config_fill(struct hostapd_config *conf,
 			wpa_printf(MSG_WARNING, "Line %d: greylist_logpath: greylist feature is not implemented yet", line);
 	} else if (os_strcmp(buf, "greylist_file") == 0) {
 			wpa_printf(MSG_WARNING, "Line %d: greylist_file: greylist feature is not implemented yet", line);
+	} else if (os_strcmp(buf, "whm_config_file") == 0) {
+		wpa_printf(MSG_WARNING, "Line %d: whm_config_file: WHM feature is not implemented yet", line);
 #ifdef CONFIG_VENDOR_MXL
 	} else if (mxl_hostapd_config_fill(conf, bss, buf, pos, line)) {
 #endif /* CONFIG_VENDOR_MXL */
-- 
2.43.0

