From 5b7906a6bb557b07af21c7dc32edda0550117f54 Mon Sep 17 00:00:00 2001
From: rportnikh <rportnikh@maxlinear.com>
Date: Sun, 26 Nov 2023 17:17:29 +0200
Subject: [PATCH] WLANRTSYS-60531- change wpa-ctrl event to per vap instead of
 per radio

---
 hostapd/ctrl_iface.c | 11 ++++++++---
 1 file changed, 8 insertions(+), 3 deletions(-)

diff --git a/hostapd/ctrl_iface.c b/hostapd/ctrl_iface.c
index e99448e6f..ab16898b1 100644
--- a/hostapd/ctrl_iface.c
+++ b/hostapd/ctrl_iface.c
@@ -6994,13 +6994,17 @@ static void hostapd_ctrl_iface_msg_cb(void *ctx, int level,
 				      const char *txt, size_t len)
 {
 	struct hostapd_data *hapd = ctx;
-	char *buf, *pos;
-	size_t total, ifacelen, bufsize;
-	errno_t err;
 
 	if ((hapd == NULL) || (hapd->iface->bss[0] == NULL))
 		return;
 
+#ifndef CONFIG_CTRL_IFACE_PER_RADIO
+	hostapd_ctrl_iface_send(hapd, level, type, txt, len);
+#else
+	char *buf, *pos;
+	size_t total, ifacelen, bufsize;
+	errno_t err;
+
 	pos = os_strchr(txt, ' ');
 	if (pos == NULL)
 		total = len;
@@ -7050,6 +7054,7 @@ static void hostapd_ctrl_iface_msg_cb(void *ctx, int level,
 	hapd = hapd->iface->bss[0];
 	hostapd_ctrl_iface_send(hapd, level, type, buf, len + ifacelen + 1);
 	os_free(buf);
+#endif /* CONFIG_CTRL_IFACE_PER_RADIO */
 }
 
 
-- 
2.43.0

