From 6cf48416fc9f792ba69526c38577406a4efe064e Mon Sep 17 00:00:00 2001
From: pvaish <pvaish@maxlinear.com>
Date: Mon, 12 Jan 2026 03:01:59 -0800
Subject: [PATCH] WLANRTSYS-99479 Fixing the RNR for 2GHz and 5GHz Radio.

Signed-off-by: pvaish <pvaish@maxlinear.com>
---
 src/ap/hostapd.c | 18 ++++++------------
 1 file changed, 6 insertions(+), 12 deletions(-)

diff --git a/src/ap/hostapd.c b/src/ap/hostapd.c
index e98523b6d..ea7e35768 100644
--- a/src/ap/hostapd.c
+++ b/src/ap/hostapd.c
@@ -299,11 +299,7 @@ int hostapd_reload_config(struct hostapd_iface *iface)
 		for (j = 0; j < iface->num_bss; j++)
 			hostapd_reload_bss(iface->bss[j]);
 #ifdef CONFIG_VENDOR_MXL
-		if (is_6ghz_op_class(hapd->iconf->op_class)) {
-			return mxl_update_rnr_check(iface);
-		}
-
-
+		return mxl_update_rnr_check(iface);
 #endif /* CONFIG_VENDOR_MXL */
 		return 0;
 	}
@@ -423,13 +419,11 @@ int hostapd_reload_config(struct hostapd_iface *iface)
 	hostapd_config_free(oldconf);
 
 #ifdef CONFIG_VENDOR_MXL
-		if (is_6ghz_op_class(iface->conf->op_class)) {
-			if (mxl_update_rnr_check(iface)) {
-				wpa_printf(MSG_ERROR,
-					   "Failed to update RNR after config reload");
-				return -1;
-			}
-		}
+	if (mxl_update_rnr_check(iface)) {
+		wpa_printf(MSG_ERROR,
+					 "Failed to update RNR after config reload");
+		return -1;
+	}
 #endif /* CONFIG_VENDOR_MXL */
 
 	return 0;
-- 
2.43.0

