From 65e9b2a6a7fe377a4768e41dabd1b108722f4c7e Mon Sep 17 00:00:00 2001
From: Ganesh Miryala <mganesh@maxlinear.com>
Date: Fri, 23 Jan 2026 17:11:37 +0530
Subject: [PATCH] SWDEVOSP-864 Don't merge the changes

---
 afc/afcd.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/afc/afcd.c b/afc/afcd.c
index b103988c7..b1d371bfa 100644
--- a/afc/afcd.c
+++ b/afc/afcd.c
@@ -5,6 +5,7 @@
  * This software may be distributed under the terms of the BSD license.
  * See README for more details.
  */
+ 
 
 #ifndef CONFIG_VENDOR_MXL /* Should only trigger in afcd build */
 #define CONFIG_VENDOR_MXL
-- 
2.43.0

