From 8c634a77dbdcc6ed213a0eb0961db21a9c182698 Mon Sep 17 00:00:00 2001
From: Veerendranath Jakkam <quic_vjakkam@quicinc.com>
Date: Wed, 7 Aug 2024 14:36:30 +0530
Subject: [PATCH] RSNO: Add debug prints for RSN override elements in EAPOL
 frames

Signed-off-by: Veerendranath Jakkam <quic_vjakkam@quicinc.com>
---
 src/common/wpa_common.c | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/src/common/wpa_common.c b/src/common/wpa_common.c
index 6e97972a2..cf6e76ef3 100644
--- a/src/common/wpa_common.c
+++ b/src/common/wpa_common.c
@@ -3752,24 +3752,36 @@ static int wpa_parse_generic(const u8 *pos, struct wpa_eapol_ie_parse *ie)
 	if (selector == RSNE_OVERRIDE_IE_VENDOR_TYPE) {
 		ie->rsne_override = pos;
 		ie->rsne_override_len = dlen;
+		wpa_hexdump(MSG_DEBUG,
+			    "RSN: RSNE Override element in EAPOL-Key",
+			    ie->rsne_override, ie->rsne_override_len);
 		return 0;
 	}
 
 	if (selector == RSNE_OVERRIDE_2_IE_VENDOR_TYPE) {
 		ie->rsne_override_2 = pos;
 		ie->rsne_override_2_len = dlen;
+		wpa_hexdump(MSG_DEBUG,
+			    "RSN: RSNE Override 2 element in EAPOL-Key",
+			    ie->rsne_override_2, ie->rsne_override_2_len);
 		return 0;
 	}
 
 	if (selector == RSNXE_OVERRIDE_IE_VENDOR_TYPE) {
 		ie->rsnxe_override = pos;
 		ie->rsnxe_override_len = dlen;
+		wpa_hexdump(MSG_DEBUG,
+			    "RSN: RSNXE Override element in EAPOL-Key",
+			    ie->rsnxe_override, ie->rsnxe_override_len);
 		return 0;
 	}
 
 	if (selector == RSN_SELECTION_IE_VENDOR_TYPE) {
 		ie->rsn_selection = p;
 		ie->rsn_selection_len = left;
+		wpa_hexdump(MSG_DEBUG,
+			    "RSN: RSN Selection element in EAPOL-Key",
+			    ie->rsn_selection, ie->rsn_selection_len);
 		return 0;
 	}
 
-- 
2.43.0

