From 460df51ed891eee6a829b4bbfd656575e0e4d581 Mon Sep 17 00:00:00 2001
From: Jouni Malinen <quic_jouni@quicinc.com>
Date: Mon, 4 Mar 2024 18:48:06 +0200
Subject: [PATCH] tests: Verify supported operating classes indication for 6
 GHz

Signed-off-by: Jouni Malinen <quic_jouni@quicinc.com>
---
 tests/hwsim/test_eht.py | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/tests/hwsim/test_eht.py b/tests/hwsim/test_eht.py
index bab67e9e5..18ade87e3 100644
--- a/tests/hwsim/test_eht.py
+++ b/tests/hwsim/test_eht.py
@@ -4,6 +4,7 @@
 # This software may be distributed under the terms of the BSD license.
 # See README for more details.
 
+import binascii
 import hostapd
 from utils import *
 from hwsim import HWSimRadio
@@ -1329,6 +1330,12 @@ def _test_eht_6ghz(dev, apdev, channel, op_class, ccfs1):
 
         eht_verify_status(dev[0], hapd, freq, bw)
         eht_verify_wifi_version(dev[0])
+        sta = hapd.get_sta(dev[0].own_addr())
+        if 'supp_op_classes' not in sta:
+            raise Exception("supp_op_classes not indicated")
+        supp_op_classes = binascii.unhexlify(sta['supp_op_classes'])
+        if op_class not in supp_op_classes:
+            raise Exception("STA did not indicate support for opclass %d" % op_class)
         hwsim_utils.test_connectivity(dev[0], hapd)
         dev[0].request("DISCONNECT")
         dev[0].wait_disconnected()
-- 
2.43.0

