From d5d0012bfb5ff25a1ddb3910386ad368e8666186 Mon Sep 17 00:00:00 2001
From: Jouni Malinen <quic_jouni@quicinc.com>
Date: Thu, 11 Jul 2024 23:16:05 +0300
Subject: [PATCH] tests: Verify bigtk_set=1 indication

Signed-off-by: Jouni Malinen <quic_jouni@quicinc.com>
---
 tests/hwsim/test_ap_pmf.py | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/tests/hwsim/test_ap_pmf.py b/tests/hwsim/test_ap_pmf.py
index 4b2cc5bbd..81782ebe2 100644
--- a/tests/hwsim/test_ap_pmf.py
+++ b/tests/hwsim/test_ap_pmf.py
@@ -1426,10 +1426,14 @@ def run_ap_pmf_beacon_protection(dev, apdev, cipher):
     # STA with Beacon protection enabled
     dev[0].connect(ssid, psk="12345678", ieee80211w="2", beacon_prot="1",
                    key_mgmt="WPA-PSK-SHA256", proto="WPA2", scan_freq="2412")
+    if dev[0].get_status_field("bigtk_set") != "1":
+        raise Exception("bigtk_set=1 not indicated")
 
     # STA with Beacon protection disabled
     dev[1].connect(ssid, psk="12345678", ieee80211w="2",
                    key_mgmt="WPA-PSK-SHA256", proto="WPA2", scan_freq="2412")
+    if dev[1].get_status_field("bigtk_set") == "1":
+        raise Exception("Unexpected bigtk_set=1 indication")
 
     time.sleep(1)
     check_mac80211_bigtk(dev[0], hapd)
-- 
2.43.0

