From be2d195843f53f64ba17513b31351b9491128944 Mon Sep 17 00:00:00 2001
From: Jouni Malinen <j@w1.fi>
Date: Sun, 22 Dec 2024 23:40:05 +0200
Subject: [PATCH] tests: Remove setting of the now removed
 p2p_pairing_verification

Signed-off-by: Jouni Malinen <j@w1.fi>
---
 tests/hwsim/test_p2p2.py | 1 -
 1 file changed, 1 deletion(-)

diff --git a/tests/hwsim/test_p2p2.py b/tests/hwsim/test_p2p2.py
index 5410e14c6..155bd5239 100644
--- a/tests/hwsim/test_p2p2.py
+++ b/tests/hwsim/test_p2p2.py
@@ -24,7 +24,6 @@ def set_p2p2_configs(dev):
     dev.request("P2P_SET supported_bootstrapmethods 6")
     dev.request("P2P_SET pairing_setup 1")
     dev.request("P2P_SET pairing_cache 1")
-    dev.request("P2P_SET pairing_verification 1")
 
 def test_p2p_usd_publish_invalid_param(dev):
     """P2P USD Publish with invalid parameters"""
-- 
2.43.0

