From eeb4493c53cbbbb95cea256fd827e435debafcc5 Mon Sep 17 00:00:00 2001
From: Vinay Gannevaram <quic_vganneva@quicinc.com>
Date: Sat, 11 Jan 2025 12:14:34 +0530
Subject: [PATCH] tests: Update P2P_INVITE parameters for P2P2 reinvoke

Signed-off-by: Vinay Gannevaram <quic_vganneva@quicinc.com>
---
 tests/hwsim/test_p2p2.py | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/tests/hwsim/test_p2p2.py b/tests/hwsim/test_p2p2.py
index f5f6af9a0..68bd1406d 100644
--- a/tests/hwsim/test_p2p2.py
+++ b/tests/hwsim/test_p2p2.py
@@ -543,7 +543,7 @@ def test_p2p_pairing_verification(dev, apdev):
     peer = wpas.get_peer(dev[1].p2p_dev_addr())
     if 'persistent' not in peer:
         raise Exception("Missing information on persistent group for the peer")
-    cmd = "P2P_INVITE persistent=" + peer['persistent'] + " peer=" + dev[1].p2p_dev_addr() + " p2p2"
+    cmd = "P2P_INVITE persistent peer=" + dev[1].p2p_dev_addr() + " p2p2"
     id0 = wpas.global_request(cmd)
     if "FAIL" in id0:
         raise Exception("P2P_INVITE Failed")
-- 
2.43.0

