From 2ddc499816ec9bf7146d30a29238dd37e8ba00db Mon Sep 17 00:00:00 2001
From: Vinay Gannevaram <quic_vganneva@quicinc.com>
Date: Thu, 20 Feb 2025 16:13:26 +0530
Subject: [PATCH] tests: Add capability checks to P2P2 test cases

Signed-off-by: Vinay Gannevaram <quic_vganneva@quicinc.com>
---
 tests/hwsim/test_p2p2.py | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/tests/hwsim/test_p2p2.py b/tests/hwsim/test_p2p2.py
index 17315cbe8..582330443 100644
--- a/tests/hwsim/test_p2p2.py
+++ b/tests/hwsim/test_p2p2.py
@@ -21,6 +21,12 @@ from test_pasn import check_pasn_capab
 def check_p2p2_capab(dev):
     check_nan_usd_capab(dev)
     check_pasn_capab(dev)
+    res = dev.request("GET_CAPABILITY p2p2")
+    if res is None or "not supported" in res:
+        raise HwsimSkip("P2P2 not supported")
+    res = dev.request("GET_CAPABILITY pcc_mode")
+    if res is None or "not supported" in res:
+        raise HwsimSkip("PCC mode not supported")
 
 def set_p2p2_configs(dev):
     dev.global_request("P2P_SET pasn_type 3")
-- 
2.43.0

