From b2455d2d2a559ae321b11618450365e018915dc3 Mon Sep 17 00:00:00 2001
From: fromans <fromans@maxlinear.com>
Date: Fri, 17 Nov 2023 12:46:12 +0200
Subject: [PATCH] Temporary: fix 32 bit compilation issues

---
 src/ap/ieee802_11_eht.c | 17 ++++++++++++-----
 1 file changed, 12 insertions(+), 5 deletions(-)

diff --git a/src/ap/ieee802_11_eht.c b/src/ap/ieee802_11_eht.c
index 577fa563b..8e0a7d3d1 100644
--- a/src/ap/ieee802_11_eht.c
+++ b/src/ap/ieee802_11_eht.c
@@ -1222,7 +1222,7 @@ u16 hostapd_process_ml_assoc_req(struct hostapd_data *hapd,
 		wpa_printf(MSG_DEBUG, "MLD: EXT ML capabilities not present");
 	}
 
-	wpa_printf(MSG_DEBUG, "MLD: expected_common_info_len=%lu",
+	wpa_printf(MSG_DEBUG, "MLD: expected_common_info_len=%zu",
 		   common_info_len);
 
 	if (sizeof(*ml) + common_info_len > ml_len) {
@@ -1392,10 +1392,17 @@ u16 hostapd_process_ml_assoc_req(struct hostapd_data *hapd,
 
 		pos = sub_elem_end;
 
-		wpa_printf(MSG_DEBUG, "MLD: link ctrl=0x%x, " MACSTR
-			   ", nstr bitmap len=%u",
-			   control, MAC2STR(link_info->peer_addr),
-			   link_info->nstr_bitmap_len);
+#ifdef CONFIG_VENDOR_MXL
+      wpa_printf(MSG_DEBUG, "MLD: link ctrl=0x%x, " MACSTR
+            ", nstr bitmap len=%zu",
+            control, MAC2STR(link_info->peer_addr),
+            link_info->nstr_bitmap_len);
+#else
+      wpa_printf(MSG_DEBUG, "MLD: link ctrl=0x%x, " MACSTR
+            ", nstr bitmap len=%lu",
+            control, MAC2STR(link_info->peer_addr),
+            link_info->nstr_bitmap_len);
+#endif /* CONFIG_VENDOR_MXL */
 
 		link_info->valid = true;
 	}
-- 
2.43.0

